OpenRA.Mods.Common/EditorBrushes/EditorBlit.cs

```diff
@@ -211,42 +222,63 @@ public static IEnumerable<IRenderable> PreviewBlitSource(
 			EditorBlitSource blitSource,
 			MapBlitFilters filters,
 			CVec offset,
-			WorldRenderer wr)
+			WorldRenderer wr,
+			bool stickToGround)
 		{
 			var world = wr.World;
 			var map = world.Map;
-
-			var terrainRenderer = world.WorldActor.Trait<ITiledTerrainRenderer>();
-			var resourceRenderers = world.WorldActor.TraitsImplementing<IResourceRenderer>().ToArray();
-
-			var wOffset = map.CenterOfCell(CPos.Zero + offset) - map.CenterOfCell(CPos.Zero);
+			var mapHeight = map.Height;
+			var mapGrid = map.Grid;
 
 			if (filters.HasFlag(MapBlitFilters.Terrain))
 			{
-				foreach (var (cpos, tile) in blitSource.Tiles)
+				var terrainRenderer = world.WorldActor.Trait<ITiledTerrainRenderer>();
+				foreach (var (pos, tile) in blitSource.Tiles)
 				{
-					var preview =
-						terrainRenderer.RenderPreview(
-							wr,
-							tile.TerrainTile,
-							map.CenterOfCell(cpos + offset));
+					var cPos = pos + offset;
+					var height = stickToGround ? (mapHeight.TryGetValue(cPos, out var isoHeight) ? isoHeight : byte.MinValue) : tile.Height;
+					var wPos = CellLayerUtils.CPosToWPos(cPos, height, mapGrid.Type);
+					var preview = terrainRenderer.RenderPreview(wr, tile.TerrainTile, wPos);
+
 					foreach (var renderable in preview)
 						yield return renderable;
 				}
 			}
 
 			if (filters.HasFlag(MapBlitFilters.Resources))
 			{
-				foreach (var (cpos, tile) in blitSource.Tiles)
+				var resourceRenderers = world.WorldActor.TraitsImplementing<IResourceRenderer>().ToArray();
+				var resourceLayer = world.WorldActor.Trait<IResourceLayer>();
+				foreach (var (pos, tile) in blitSource.Tiles)
 				{
 					if (tile.ResourceLayerContents == null || tile.ResourceLayerContents.Value.Type == null)
 						continue;
 
+					var cPos = pos + offset;
+					if (!filters.HasFlag(MapBlitFilters.Terrain) && !resourceLayer.CanAddResource(tile.ResourceLayerContents.Value.Type, cPos))
+						continue;
+
+					byte height;
+					if (stickToGround)
+					{
+						if (!mapHeight.TryGetValue(cPos, out height))
+							height = byte.MinValue;
+
+						if (filters.HasFlag(MapBlitFilters.Terrain))
+							height += map.Rules.TerrainInfo.GetTerrainInfo(tile.TerrainTile).Height;
+					}
+					else
+					{
+						if (filters.HasFlag(MapBlitFilters.Terrain))
+							height = tile.Height;
+						else if (!mapHeight.TryGetValue(cPos, out height))
+							height = byte.MinValue;
+					}
```

Comment:
Arguably slightly simpler, albeit perhaps slightly less optimized:

```suggestion
					if (!mapHeight.TryGetValue(cPos, out var height))
						height = byte.MinValue;

					if (filters.HasFlag(MapBlitFilters.Terrain))
					{
						if (stickToGround)
							height += map.Rules.TerrainInfo.GetTerrainInfo(tile.TerrainTile).Height;
						else
							height = tile.Height;
					}
```

Reply:
I think the current version better represents intent

Reply:
Adjusted to even better represent intent